### DISCUSSION TITLE
Reactive components are being reloaded unnecessarily

### DISCUSSION THREAD

**Discussion by alexwass-lr** [livewire/livewire#8186] [7. Bugs (archived)] [answered]
### Livewire version

v3.4.9

### Laravel version

v10.47.0

### Which PHP version are you using?

PHP 8.1

### Steps To Reproduce

1. Create two Livewire components, one for the parent and the other the children
2. Within the parent, create a loop to generate a list, rendering the child component
3. In the parent, provide a way to filter the nested children (e.g. search)
4. Pass a reactive property to each child component from the parent

### Repository or Wirebox URL

https://wirebox.app/b/x6drg

### Additional information

I've been trying to troubleshoot a slow section in on one of our sites - it's effectively a filter for a table. There is an overarching parent component, which loads a child component for each row. There is a `Reactive` property that is passed to the child to help calculate some computed values based on some other table filters. We noticed the main issue was the first search after page load, it was actually slower than page load because all the child components are `Lazy` loaded. Then more confusingly when refreshing the filtered page it loads much quicker (_we persist this property with the `Url` attribute_) .

After some digging it seems the dehydration for these child components aren't performing as I'd expect, and is actually loading all the initial child components again (_and not lazy, hence the performance drop_), even if they've been excluded by the search. Specifically commenting out this line adjusts the responses to how I'd expect: https://github.com/livewire/livewire/blob/main/src/Features/SupportReactiveProps/BaseReactive.php#L43

Hope that makes some sense...if not, hopefully the below demo will help!

Taking the [Wirebox](https://wirebox.app/b/x6drg) example, (_you'll need the network dev tools open for this_) when you do an initial search for `app`, you can see each component from the initial load has been requested as part of the response:

```
{
    "components": [
        {
            "snapshot": "{\"data\":{\"search\":\"app\"},\"memo\":{\"id\":\"PmEUc0J38GRYQrqnewVz\",\"name\":\"search\",\"path\":\"\\\/\",\"method\":\"GET\",\"children\":{\"Apple\":[\"li\",\"VqR9whxWv7gzVW4hrp99\"]},\"scripts\":[],\"assets\":[],\"errors\":[],\"locale\":\"en\"},\"checksum\":\"7e5b9eff37adbd1c9f8687d350ffc0a7eff788da24d8a810c3bfa83691b0e388\"}",
            "effects": {
                "returns": [],
                "html": "<div wire:id=\"PmEUc0J38GRYQrqnewVz\" class=\"bg-white h-full\">\n    <input type=\"text\" wire:model.live.debounce.500ms=\"search\" placeholder=\"Search\">\n\n    <ul>\n        <!--[if BLOCK]><![endif]-->            <li wire:id=\"VqR9whxWv7gzVW4hrp99\"><\/li>        <!--[if ENDBLOCK]><![endif]-->\n    <\/ul>\n<\/div>"
            }
        },
        {
            "snapshot": "{\"data\":{\"something\":\"test\",\"fruit\":\"Apple\"},\"memo\":{\"id\":\"VqR9whxWv7gzVW4hrp99\",\"name\":\"fruit\",\"path\":\"\\\/\",\"method\":\"GET\",\"children\":[],\"scripts\":[],\"assets\":[],\"props\":[\"something\"],\"errors\":[],\"locale\":\"en\"},\"checksum\":\"505569c89a5e50080d7a31c208f0ae9af7fa49ca681093650bf6e6b5d81ac9da\"}",
            "effects": {
                "returns": [],
                "html": "<li wire:id=\"VqR9whxWv7gzVW4hrp99\">Apple<\/li>"
            }
        },
        {
            "snapshot": "{\"data\":{\"something\":\"test\",\"fruit\":\"Banana\"},\"memo\":{\"id\":\"pDgQFG9uMZcLWq0huomI\",\"name\":\"fruit\",\"path\":\"\\\/\",\"method\":\"GET\",\"children\":[],\"scripts\":[],\"assets\":[],\"props\":[\"something\"],\"errors\":[],\"locale\":\"en\"},\"checksum\":\"c2af25fcdefe713454ddc4bf9871209b8d7b085fbf50a47751d630bcaa1a4233\"}",
            "effects": {
                "returns": [],
                "html": "<li wire:id=\"pDgQFG9uMZcLWq0huomI\">Banana<\/li>"
            }
        },
        {
            "snapshot": "{\"data\":{\"something\":\"test\",\"fruit\":\"Cherry\"},\"memo\":{\"id\":\"MX9zlXjMZ993MLx6ArRQ\",\"name\":\"fruit\",\"path\":\"\\\/\",\"method\":\"GET\",\"children\":[],\"scripts\":[],\"assets\":[],\"props\":[\"something\"],\"errors\":[],\"locale\":\"en\"},\"checksum\":\"7555896dde927b22fb042dd5a8a7ad4ae5d7f6769b307f9f22dde8beeab62d2a\"}",
            "effects": {
                "returns": [],
                "html": "<li wire:id=\"MX9zlXjMZ993MLx6ArRQ\">Cherry<\/li>"
            }
        },
        {
            "snapshot": "{\"data\":{\"something\":\"test\",\"fruit\":\"Date\"},\"memo\":{\"id\":\"4Ff9KbFlJihf4BHQKXL0\",\"name\":\"fruit\",\"path\":\"\\\/\",\"method\":\"GET\",\"children\":[],\"scripts\":[],\"assets\":[],\"props\":[\"something\"],\"errors\":[],\"locale\":\"en\"},\"checksum\":\"809c6614b55583f9858f1f4c6d0e15b8ee517190f8783b5dc863a10229b60bea\"}",
            "effects": {
                "returns": [],
                "html": "<li wire:id=\"4Ff9KbFlJihf4BHQKXL0\">Date<\/li>"
            }
        },
        {
            "snapshot": "{\"data\":{\"something\":\"test\",\"fruit\":\"Elderberry\"},\"memo\":{\"id\":\"gAYdDUZq6uW5hAFdKh9d\",\"name\":\"fruit\",\"path\":\"\\\/\",\"method\":\"GET\",\"children\":[],\"scripts\":[],\"assets\":[],\"props\":[\"something\"],\"errors\":[],\"locale\":\"en\"},\"checksum\":\"e78a31281086d598253033e9d9248f8201538e3ac6225d0232334437ca0d8419\"}",
            "effects": {
                "returns": [],
                "html": "<li wire:id=\"gAYdDUZq6uW5hAFdKh9d\">Elderberry<\/li>"
            }
        },
        {
            "snapshot": "{\"data\":{\"something\":\"test\",\"fruit\":\"Fig\"},\"memo\":{\"id\":\"CP6ZeKx4UvcoLTHyY2pW\",\"name\":\"fruit\",\"path\":\"\\\/\",\"method\":\"GET\",\"children\":[],\"scripts\":[],\"assets\":[],\"props\":[\"something\"],\"errors\":[],\"locale\":\"en\"},\"checksum\":\"96076d4859b20e5a69eb1688fb5cd65fc6f131f3d68dd8b421ec0751513450bf\"}",
            "effects": {
                "returns": [],
                "html": "<li wire:id=\"CP6ZeKx4UvcoLTHyY2pW\">Fig<\/li>"
            }
        }
    ],
    "assets": []
}
```

It seems to be one state behind, because if you search again, the response will contain the previous result. As you can see, we're not updating the `Reactive` property `$something`, so I wouldn't expect those components to be reloaded.

Now if you remove the `#[Reactive]` attribute from `Fruit::class`, this is the response we get instead:

```
{
    "components": [
        {
            "snapshot": "{\"data\":{\"search\":\"app\"},\"memo\":{\"id\":\"tiartsSfltwuYhBCG4Uq\",\"name\":\"search\",\"path\":\"\\\/\",\"method\":\"GET\",\"children\":{\"Apple\":[\"li\",\"jgl0GadRUVlXQmbIr3zO\"]},\"scripts\":[],\"assets\":[],\"errors\":[],\"locale\":\"en\"},\"checksum\":\"4a9fbb87fe95d0a036a73fe80a8a996aa41071637b8eebabfe0decb6d8613640\"}",
            "effects": {
                "returns": [],
                "html": "<div wire:id=\"tiartsSfltwuYhBCG4Uq\" class=\"bg-white h-full\">\n    <input type=\"text\" wire:model.live.debounce.500ms=\"search\" placeholder=\"Search\">\n\n    <ul>\n        <!--[if BLOCK]><![endif]-->            <li wire:id=\"jgl0GadRUVlXQmbIr3zO\"><\/li>        <!--[if ENDBLOCK]><![endif]-->\n    <\/ul>\n<\/div>"
            }
        }
    ],
    "assets": []
}
```

Obviously this isn't a major problem for this demo, but some of our component rows can take over a second to finish their calculations (_hence the use of Lazy_) so I don't want them running again if someone is searching for something else. 

Am I missing something here and this would be considered expected behaviour @joshhanley? I'm happy to dig deeper and PR some tests/fix if it's not though.

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by joshhanley** [comment]
@alexwass-lr hmm struggled to follow what your explaining but I think I get it.

Basically you are saying, that when the request is returned after searching for "app", the response should only contain the parent and the "apple" component but it instead has all of them.

If that is correct then yeah that is unexpected in my opinion. So I would say yeah if you can open failing test PR, that'd be great!

**Reply by alexwass-lr** [reply]
@joshhanley sorry I did try not to waffle 😅

In summary, exactly that. To clarify, all the returned components are not being rendered though, but they still get computed which is causing the loading delay.

I'll try and take a look at the failing PR this week - thanks!

**Reply by alexwass-lr** [reply]
I've been trying to find the best way to test this @joshhanley but I seem to be running into the same issue [this skipped test](https://github.com/livewire/livewire/blob/main/src/Features/SupportReactiveProps/UnitTest.php#L31) does, it throws the error `Unable to find component` when trying to update component properties with children.

However, during my digging to create the test, I've gained a better understanding of the cause:
1. During [dehydration](https://github.com/livewire/livewire/blob/main/src/Features/SupportReactiveProps/BaseReactive.php#L43), the attribute adds the `props` key to the component
2. On the next request, when the commit is built, it uses this key to [find all reactive child components](https://github.com/livewire/livewire/blob/main/js/features/supportPropsAndModelables.js#L9) to include with the request
3. When the request is handled, it then [loops through all components](https://github.com/livewire/livewire/blob/main/src/Mechanisms/HandleRequests/HandleRequests.php#L84), which are the parent and all active children
4. This then triggers each component to be updated, [along with their properties](https://github.com/livewire/livewire/blob/main/src/Mechanisms/HandleComponents/HandleComponents.php#L97) and any other events

Whilst this is ideal for normal reactive updates, when the parent has a property updated which changes the children, the destroyed children should not be updated - this is the performance issue I'm running into.

In my opinion, I would expect the destroyed children to be removed before any reactive updates are applied. There are several places I can think to do this, but slightly break the isolation of the feature folder structure. 

What are your thoughts on this?

**Reply by alexwass-lr** [reply, marked as answer]
I ended up creating a draft pull request for this @joshhanley to demonstrate one of my suggested solutions: https://github.com/livewire/livewire/pull/8288

Again, not sure if this is the best solution, or how this would be testable, but it seems to fix the problem in my scenario.

Would appreciate your feedback when you get a moment!
